### PR TITLE
Add weight_predictions function

### PR DESCRIPTION
**Pull request by aloctavodia**
This function will take a list of idatas with `posterior_predictive` groups and a list of model weights (computed using `az.compare` or something else) and it will return a new inference data with a `posterior_predictive` group composed of weighted samples from the input idatas. 

@zaxtax maybe we can focus on this and forgot about `pm.sample_posterior_predictive_w`

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format
- [x] Includes new or updated tests to cover the new feature
- [x] Code style  correct (follows pylint and black guidelines)
- [x] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#v0xx-unreleased)

----
:books: Documentation preview :books:: https://arviz--2147.org.readthedocs.build/en/2147/

### CODE REVIEW

**Comment by zaxtax** [comment]
I think this is a fantastic idea. I always found
`pm.sample_posterior_predictive_w` awkward.

One check that I think we need is something ensuring all models have
identical observed data.

**Comment by zaxtax** [comment]
Should we just remove sample_posterior_predictive_w or throw an error
pointing people to this function?

**Comment by aloctavodia** [comment]
glad you like it. yeah we are assuming a few things here, we should add some check. We may want to deprecate `pm.sample_posterior_predictive_w` and point people here. and also update/improve the example in PyMC-examples

File: arviz/stats/stats.py [hunk 1]
```diff
@@ -2043,3 +2044,61 @@ def apply_test_function(
...
+    idata: InferenceData
+        Output InferenceData object with the groups `posterior_predictive` and `observed_data`.
+
+    See Also
+    --------
+    compare :  Compare models based on PSIS-LOO `loo` or WAIC `waic` cross-validation
+    """
+    if weights is None:
+        weights = np.ones(len(idatas)) / len(idatas)
+    weights = np.array(weights, dtype=float)
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2076]:
```suggestion
    weights = np.asarray(weights, dtype=float)
```

File: arviz/stats/stats.py [hunk 2]
```diff
@@ -2043,3 +2044,61 @@ def apply_test_function(
...
+    weights /= weights.sum()
+
+    if not np.all(["posterior_predictive" in idata.groups() for idata in idatas]):
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2079]:
```suggestion
    if not all("posterior_predictive" in idata.groups() for idata in idatas):
```

File: arviz/stats/stats.py [hunk 3]
```diff
@@ -2043,3 +2044,61 @@ def apply_test_function(
...
+        raise ValueError(
+            "All the InferenceData objects must contain the `posterior_predictive` group"
+        )
+
+    if not np.all([idatas[0].observed_data.equals(idata.observed_data) for idata in idatas[1:]]):
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2084]:
```suggestion
    if not all(idatas[0].observed_data.equals(idata.observed_data) for idata in idatas[1:]):
```

Also need to handle cases with 1 or 0 idatas len

**Review by ahartikainen — COMMENTED** [review]
Couple of comments

**Comment by aloctavodia** [comment]
ready for review or merge, depending on how benevolent you are feeling today

File: arviz/stats/stats.py [hunk 4]
```diff
@@ -2043,3 +2044,69 @@ def apply_test_function(
...
+
+
+def weight_predictions(idatas, weights=None):
+    """
+    Generate weighted posterior predictive samples from a list of InferenceData
+    and a set of weights.
+
+    Parameters
+    ---------
+    idatas : list[InfereneData]
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2056]:
```suggestion
    idatas : list[InferenceData]
```

File: arviz/stats/stats.py [hunk 5]
```diff
@@ -2043,3 +2044,69 @@ def apply_test_function(
...
+
+    if weights is None:
+        weights = np.ones(len(idatas)) / len(idatas)
+    elif len(idatas) != len(weights):
+        raise ValueError(
+            "The number of weights should be the same as the number of InferenceData objects"
+        )
+
+    weights = np.asarray(weights, dtype=float)
+    weights /= weights.sum()
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2093]: Does this step copy weights?

File: arviz/stats/stats.py [hunk 6]
```diff
@@ -2043,3 +2044,69 @@ def apply_test_function(
...
+
+    len_idatas = [
+        len(idata.posterior_predictive.chain) * len(idata.posterior_predictive.draw)
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2096]: You could use dim info here?

File: arviz/stats/stats.py [hunk 7]
```diff
@@ -2043,3 +2044,69 @@ def apply_test_function(
...
+        for idata in idatas
+    ]
+
+    new_samples = (np.min(len_idatas) * weights).astype(int)
+
```
Comment (ahartikainen) [inline arviz/stats/stats.py:2101]: Could all samples go to 0?

**Review by ahartikainen — APPROVED** [review]

**Comment by oussamadhaoui** [comment]
`ppc_w= az.stats.weight_predictions([model0, model1, model2], weights)`
I used this line and I received this error
`ValueError: All the InferenceData objects must contain the `posterior_predictive` group
`

**Comment by aloctavodia** [comment]
Hi @oussamadhaoui,

 did you check that `model0`, `model1`, and `model2` have the `posterior_predictive` group? 
 
 also for the future, it is usually the best idea to open a new issue ticket instead of adding comments to merged PRs. Check this [guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue)

**Pull request merged**